crypto/openssl_csr.py

```diff
@@ -0,0 +1,299 @@
+        }
+
+        if self.subjectAltName is None:
+            self.subjectAltName = 'DNS:%s' % self.subject['CN']
+
+        for (key,value) in self.subject.items():
+            if value is None:
+                del self.subject[key]
+         
+
+    @property
+    def file_path(self):
+        return self.path
+   
+
+    def generate(self, module):
+        '''Generate the certificate signing request.'''
+
+        if not os.path.exists(self.file_path) or self.force:
+            req = crypto.X509Req()
+            req.set_version(self.version)
+            subject = req.get_subject()
+            for (key,value) in self.subject.items():
+                if value is not None:
+                    setattr(subject, key, value)
+
+            if self.subjectAltName is not None:
+                req.add_extensions([crypto.X509Extension("subjectAltName", False, self.subjectAltName)])
+
+            privatekey_content = open(self.privatekey_path).read()
+            self.privatekey = crypto.load_privatekey(crypto.FILETYPE_PEM, privatekey_content)
+
+            req.set_pubkey(self.privatekey)
+            req.sign(self.privatekey, self.digest)
+            self.request = req
+
+            csr_file = open(self.file_path, 'w')
+            csr_file.write(crypto.dump_certificate_request(crypto.FILETYPE_PEM, self.request))
+            csr_file.close()
+        else:
+            self.changed = False
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            self.changed = True
+        
+
+    def remove(self):
+        '''Remove the Certificate Signing Request.'''
+
+        try:
+            os.remove(self.file_path)
+        except OSError:
+            self.changed = False
+
+
+    def dump(self):
+        '''Serialize the object into a dictionnary.'''
+
+        result = {
+          'csr': self.file_path,
+          'subject': self.subject,
+          'subjecAltName': self.subjectAltName,
+          'changed': self.changed
+        }
+
+        return result
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            digest=dict(default='sha256', type='str'),
+            privatekey_path=dict(require=True, type='str'),
+            version=dict(default='3', type='int'),
+            force=dict(default=False, type='bool'),
+            subjectAltName=dict(aliases=['subjectAltName'], type='str'),
+            path=dict(required=True, type='str'),
+            countryName=dict(aliases=['C'], type='str'),
+            stateOrProvinceName=dict(aliases=['ST'], type='str'),
+            localityName=dict(aliases=['L'], type='str'),
+            organizationName=dict(aliases=['O'], type='str'),
+            organizationalUnitName=dict(aliases=['OU'], type='str'),
+            commonName=dict(aliases=['CN'], type='str'),
```

Comment:
As discussed, since either commonName or subjectAltName is required, you should look at required_one_of in https://github.com/ansible/ansible-modules-core/blob/devel/system/systemd.py#L244